tensorly/tenalg/proximal.py

```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+        gradient = - AtB + tl.tenalg.multi_mode_dot(x_upd, pseudo_inverse, transpose=False) + sparse
+
+        if non_negative is True:
+            delta_x = tl.where(gradient_step * gradient < x, gradient_step * gradient, x_upd)
+        else:
+            delta_x = gradient_step * gradient
+
+        xnew = x_upd - delta_x
+        momentum = (1 + tl.sqrt(1 + 4 * momentum_old ** 2)) / 2
+        x_upd = xnew + ((momentum_old - 1) / momentum) * (xnew - x)
+        momentum_old = momentum
+        x = tl.copy(xnew)
+        norm = tl.norm(delta_x)
+        if iteration == 1:
+            norm_0 = norm
+        if norm < 0.01 * norm_0:
+            break
+    return x
+
+
+def active_set_nnls(AtA, AtB, x=None, n_iter_max=100, tol=10e-8):
+    """
+     Active set algorithm for non-negative least square solution.
+
+     Parameters
+     ----------
+     AtA: ndarray
+        Pre-computed Kronecker product of the transposed of A and A.
+     AtB: ndarray
+        Pre-computed product of the transposed of A and B.
+     x: initialized array
+        Default: None
+     n_iter_max : int
+         Maximum number of iteration
+         Default: 100
+    tol : float
+         Early stopping criterion
+
+     Returns
+     -------
+     x : Updated ndarray
+
+     Reference
+     ----------
+     [1] : Bro, R., & De Jong, S. (1997). A fast non‐negativity‐constrained
+           least squares algorithm. Journal of Chemometrics: A Journal of
+           the Chemometrics Society, 11(5), 393-401.
+     """
+    
+    if x is None:
+        x_vec = tl.zeros(tl.shape(AtA)[1])
+    else:
+        x_vec = tl.base.tensor_to_vec(x)
+
+    first = AtB
+    second = tl.dot(AtA, x_vec)
+    w = first - second
+    P = x_vec > 0
+    R = x_vec <= 0
+    s = tl.zeros(tl.shape(x_vec), **tl.context(x_vec))
+
+    for iteration in range(n_iter_max):
+        # Early stopping criterion
+        if True in R and tl.max(w[R]) <= tol:
+            # Not necessary to solve, if iteration is more than 0
+            if iteration == 0:
+                try:
+                    a = tl.tensor(tl.to_numpy(AtA)[P, :][:, P], **tl.context(AtA))
+                    b = AtB[P]
+                    c = tl.solve(a, b)
+                    indice_list = []
+                    for i in range(tl.shape(s)[0]):
+                        if P[i]:
+                            indice_list.append(i)
+                            x_vec = tl.index_update(x_vec, tl.index[int(i)], c[len(indice_list) - 1])
+                        else:
+                            x_vec = tl.index_update(x_vec, tl.index[int(i)], 0)
+                    x_vec = tl.clip(x_vec, 0, tl.max(x_vec))
+                except:
+                    x_vec = tl.copy(x_vec)
+            break
+
+        if True in R:
+            indice = tl.argmax(w)
+            P = tl.index_update(P, tl.index[indice], True)
+            R = tl.index_update(R, tl.index[indice], False)
+        # To avoid singularity error when initial x exists
```

Comment:
Is that a single occurrence, can we test for that?

Reply:
When initial x exists, we may have this error. According to the my tests, it occurs when all the P (changed, now passive set) is True. So, it happens once because when P has only true, R will have only false then iteration stops in the end.

Reply:
So we could test with tl.any?